client/src/utils/ajax.ts

```diff
@@ -81,6 +81,9 @@ async function request<T>(
   };
 
   const response = await fetch(`${base}${path}`, options);
+  if (!response.ok) {
+    throw new Error(response.statusText);
+  }
```

Comment:
```suggestion
```

Throwing an error here works, but does too much. Since we know what we're looking for (`!response.ok`) we can check for just that in `postChallenge` and emit `updateFailed` when the response is not ok.  This approach is less likely to have unforeseen side effects.

Reply:
I'm happy for the PR to go in without specifically handling `!response.ok`, the other changes are already an improvement.

We can always modify `postChallenge` in another PR.